Add system report function to collect system information on cluster

francesco-giordano opened the pull request:
### Description of changes
* Added to the integration tests common utility a function to collect and store all the relevant information about the head node and a compute fleet node. This system report can be compared performing a diff on two of them.

### Tests
* Manual test with slurm_scheduler test
* Manual and automated tests with test_disable_hyperthreading, test_ad_integration and test_efa
* All the tested tests produced the system report named ad the test (eg. systemreport_compute_node_test_ad_integration[us-west-1-c5.xlarge-alinux2-slurm-benchmarks0-MicrosoftAD-ldap-False].tar.gz)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

francesco-giordano committed:
Add get_remote_files method to remove_command_executor

francesco-giordano committed:
Add run_system_analyzer method to common utils and add system-analyzer.sh script

run_system_analyzer runs system-analyzer.sh script on head and a compute node to collect system information and copy them into output_dir
Add gathering_system_report in benchmark tests
Add Ubuntu support, kernel version and network statistics to system-information
Add support to partition in run_system_analyzer
Add run_system_analyzer to test_efa in case of OSU benchmark and p4d

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+if [ -d "$1" ]; then
+  echo "Directory $1 exists."
+else
+  echo "Dectory $1 does not exist - exiting"
```
typo

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+# Script to generate a system information report.
```
let's add a copyright with apache2 license, let's list requirements (if any) and usage example.

see for example: https://github.com/aws/aws-parallelcluster-cookbook/blob/release-3.1/cookbooks/aws-parallelcluster-install/files/default/dcv/pcluster_dcv_connect.sh

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+# Script to generate a system information report.
+# This script needs to be run by a user in sudoers and it takes an existing directory as argument
+
+set -x
```
this can be set in the shebang

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+
+set -x
+
+#Check arguments
```
minor: let's keep one space between `#` and comment

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+    fi
+}
+
+#trap 'sighdl $? $LINENO' SIGKILL SIGINT SIGTERM INT ERR
```
should you remove it?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+ethtool -S eth0 > $TEMP_DIR/network/eth0_stats
+netstat -s > $TEMP_DIR/network/netstat_stats
+
+MPI_MODULES=$(for MODULE in $(echo $MODULEPATH | tr : '\n'); do ls -1 $MODULE | grep -i mpi; done)
```
let's add a comment before each code section to clarify what we're doing/testing. or better let's define functions for specific steps.

e.g. _list_installed_packages, _list_mpi_versions, _check_network, etc.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -24,6 +26,8 @@
 
 LOGGER = logging.getLogger(__name__)
 
+SYSTEM_REPORT_SCRIPT = pathlib.Path(__file__).parent / "data/system-report.sh"
```
what about calling the script "system-analyzer" or something like this? the "report" is the result.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+
+uname -a > $TEMP_DIR/uname
+
+$PACKAGE_REPORT_CMD | sort -du > $TEMP_DIR/packages
```
let's use ${VARIABLE_NAME} everywhere.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+
+uname -a > $TEMP_DIR/uname
+
+$PACKAGE_REPORT_CMD | sort -du > $TEMP_DIR/packages
```
without pipefail if the `PACKAGE_REPORT_CMD` fails the command goes ahead with the sort command. It's not a big deal here but it's a best practice to use pipefail.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+}
+
+#trap 'sighdl $? $LINENO' SIGKILL SIGINT SIGTERM INT ERR
+trap 'sighdl $? $LINENO' SIGKILL SIGINT SIGTERM INT
```
Could you please add a comment describing what is this command for?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+#Create package report commands
+PACKAGE_REPORT_CMD=""
+OS_ID="$(grep "^ID=" /etc/os-release)"
+OS_ID="${OS_ID:3}"
```
you could do this in a single line with: `$(grep ^ID= /etc/os-release | cut -d"=" -f 2 | xargs)`

enrico-usai replied:
xargs removes the trailing quotas, in this way you can remove them from the switch case.

francesco-giordano replied:
Nice call. Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+    ubuntu)
+        PACKAGE_REPORT_CMD="dpkg-query -l"
+        ;;
+    '"amzn"' | '"centos"')
```
see previous comment about xargs

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-report.sh:
```diff
@@ -0,0 +1,104 @@
+curl -H "X-aws-ec2-metadata-token: $TOKEN" http://169.254.169.254/latest/meta-data/instance-type > $TEMP_DIR/instance-type
+curl -H "X-aws-ec2-metadata-token: $TOKEN" http://169.254.169.254/latest/user-data > $TEMP_DIR/user-data
+
+rm -fr $1/systemreport.tar.gz
```
minor: system-report.tar.gz ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
     return [table["RouteTableId"] for table in response["RouteTables"]]
+
+
+def gather_system_report(cluster, request, scheduler_commands, remote_command_executor):
```
Let's add a docstring for this public utility/function.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+    out_dir = request.config.getoption("output_dir")
+    logging.info(f"Get system information about test: {request.node.name}")
+    logging.info(f"Run command on head node to get system information about test: {request.node.name}")
+    result = remote_command_executor.run_remote_script(SYSTEM_REPORT_SCRIPT, args=["/tmp"])
```
SYSTEM_REPORT_SCRIPT --> SYSTEM_ANALYZER_SCRIPT ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+    result = remote_command_executor.run_remote_script(SYSTEM_REPORT_SCRIPT, args=["/tmp"])
+    logging.debug(f"result.failed={result.failed}")
+    logging.debug(f"result.stdout={result.stdout}")
+    logging.info(f"Copy results from remote cluster into: {out_dir}/system_report/systemreport_compute_node_{request.node.name}.tar.gz")
```
systemreport_compute_node --> should be systemreport_head_node

francesco-giordano replied:
Right. Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+def gather_system_report(cluster, request, scheduler_commands, remote_command_executor):
+    out_dir = request.config.getoption("output_dir")
+    logging.info(f"Get system information about test: {request.node.name}")
+    logging.info(f"Run command on head node to get system information about test: {request.node.name}")
```
You can just write: "Retrieve head node system information for test: ...."

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+    )
+    logging.info("Copy head node system report: successfully executed")
+
+    logging.info(f"Run command on a compute node to get system information about test: {request.node.name}")
```
Same here: 
You can just write: "Retrieve compute node system information for test: ...."

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+    logging.debug(f"result.failed={result.failed}")
+    logging.debug(f"result.stdout={result.stdout}")
+    logging.info(f"Copy results from remote cluster into: {out_dir}/system_report/systemreport_compute_node_{request.node.name}.tar.gz")
+    os.makedirs(f"{out_dir}/system_report", exist_ok=True)
```
`{out_dir}/system_report` is used in multiple places let's define an internal variable for it.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+    scheduler_commands.assert_job_succeeded(job_id)
+    logging.info(f"Copy results from remote cluster into: {out_dir}/system_report/systemreport_compute_node_{request.node.name}.tar.gz")
+    remote_command_executor.get_remote_files(
+        "/opt/parallelcluster/shared/systemreport.tar.gz",
```
same here: `/opt/parallelcluster/shared` is used in more than 1 place, let's define a variable for it, to avoid typos.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+        f"{out_dir}/system_report/systemreport_head_node_{request.node.name}.tar.gz",
+        preserve_mode=False,
+    )
+    logging.info("Copy head node system report: successfully executed")
```
you could write: "Head node system information correctly retrieved."

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,33 @@ def get_route_tables(subnet_id, region):
+        f"{out_dir}/system_report/systemreport_compute_node_{request.node.name}.tar.gz",
+        preserve_mode=False,
+    )
+    logging.info("Copy compute node system report: successfully executed")
```
same here, you could write: "Compute node system information correctly retrieved."

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/disable_hyperthreading/test_disable_hyperthreading.py:
```diff
@@ -59,6 +60,7 @@ def test_hit_disable_hyperthreading(
 
     assert_no_errors_in_logs(remote_command_executor, scheduler)
     run_benchmarks(remote_command_executor, scheduler_commands)
+    gather_system_report(cluster, request, scheduler_commands, remote_command_executor)
```
what about: run_system_analyzer ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/remote_command_executor.py:
```diff
@@ -174,3 +174,22 @@ def _copy_additional_files(self, files):
         for local_remote_path in local_remote_paths or []:
             self.__connection.put(**local_remote_path)
+
+    def get_remote_files(self, *args, **kwargs):
```
files or file?

francesco-giordano replied:
I going to leave files because it copies file-like object(directory and files).

tschuett commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,30 @@ Scheduling:
+## Collect system report
+
+In case of performance degradation detected in tests, to speed up the root cause analysis it could be useful to have a
+report on all the service, cron and information about the  . The function create
```
cluster?

francesco-giordano replied:
Hi @tschuett,
thank you for your feedback, the readme it is still work in progress; when the PR will be ready I would appreciated a feedback from you.

enrico-usai commented on tests/integration-tests/README.md:
````diff
@@ -976,3 +976,30 @@ Scheduling:
 ```
+## Collect system report
+
+In case of performance degradation detected in tests, to speed up the root cause analysis it could be useful to have a
````
could be --> is

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,30 @@ Scheduling:
+## Collect system report
+
+In case of performance degradation detected in tests, to speed up the root cause analysis it could be useful to have a
+report on all the service, cron and information about the  . The function create
```
services, cron settings

francesco-giordano replied:
Done

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,30 @@ Scheduling:
+
+In case of performance degradation detected in tests, to speed up the root cause analysis it could be useful to have a
+report on all the service, cron and information about the  . The function create
+a directory in `out_dir` called `system_report`. Every run the function creates 2 files, one for the head node and one
```
create a directory --> creates a subfolder

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,30 @@ Scheduling:
+In case of performance degradation detected in tests, to speed up the root cause analysis it could be useful to have a
+report on all the service, cron and information about the  . The function create
+a directory in `out_dir` called `system_report`. Every run the function creates 2 files, one for the head node and one
+for a compute node of the fleet. The file generated are `tar.gz`, when extracted they contain a directory structure
```
The generated file is a `tar.gz` archive. The archive contains ...

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+## Collect system report
+
+In case of performance degradation detected in tests, to speed up the root cause analysis it is useful to have a
+report on all the services, cron settings and information about the cluster tested. The function creates a subfolder
```
on all the --> of all the
the cluster tested --> the tested cluster

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+In case of performance degradation detected in tests, to speed up the root cause analysis it is useful to have a
+report on all the services, cron settings and information about the cluster tested. The function creates a subfolder
+in `out_dir` called `system_analyzer`. Every run the function creates 2 files, one for the head node and one
+for a compute node of the fleet. The file generated are `tar.gz`. The archive contains a directory structure
```
The file generated are --> the file generated is a `tar.gz` archive containing a directory...

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+report on all the services, cron settings and information about the cluster tested. The function creates a subfolder
+in `out_dir` called `system_analyzer`. Every run the function creates 2 files, one for the head node and one
+for a compute node of the fleet. The file generated are `tar.gz`. The archive contains a directory structure
+which is comparable with diff. Moreover, the function collects also some network statistics which can be inspected to
```
diff --> `diff`
the function collects also --> the function also collects

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+in `out_dir` called `system_analyzer`. Every run the function creates 2 files, one for the head node and one
+for a compute node of the fleet. The file generated are `tar.gz`. The archive contains a directory structure
+which is comparable with diff. Moreover, the function collects also some network statistics which can be inspected to
+get information about package drop and other meaningful network metrics.
```
about package drop --> about dropped packages ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+get information about package drop and other meaningful network metrics.
+
+The information collected are:
+* Recognized system id (ubuntu,amzn,centos)
```
Recognized --> we can remove this word

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+The information collected are:
+* Recognized system id (ubuntu,amzn,centos)
+* uname with kernel version
+* packages installed
```
installed packages

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+* uname with kernel version
+* packages installed
+* Services and timers active on the system
+* Scheduled command like cron, at, anacron
```
commands

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+* packages installed
+* Services and timers active on the system
+* Scheduled command like cron, at, anacron
+* The mpi packages available in the system through the modules
```
The available MPI modules?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,40 @@ Scheduling:
+* Services and timers active on the system
+* Scheduled command like cron, at, anacron
+* The mpi packages available in the system through the modules
+* The network configured and the statistics associated to those network
```
configured network

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+#!/bin/bash -eua
```
let's add `-o pipefail` too

francesco-giordano replied:
In the shebang is only allowed 1 argument: in shebang leaved only bash used set in main function

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+  curl -H "X-aws-ec2-metadata-token: $TOKEN" http://169.254.169.254/latest/user-data > "${TEMP_DIR}"/user-data
+}
+
+is_user_root ()
```
let's be consistent, if you're using the underscore prefix  for private functions  and the `function`  keyword let's use it for all of them.

enrico-usai replied:
see sighdl and body too

enrico-usai replied:
for some of them you have no space between the function name and () for some of them you have it. Let's be consistent.

francesco-giordano replied:
Thank you, fixed

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+
+function _save_imds_info() {
+  # Mics IMDS info
+  TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"`
```
let's use posix syntax for subshell `$()` vs backtick

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+
+# Become root if not
+if ! is_user_root;
+then
```
let's move then in the previous line.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+  exit 1
+fi
+
+# Become root if not
```
Let's add a full explanation:
\# Become root if not and re-execute the script itself

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+if ! is_user_root;
+then
+  echo "Running the script as root"
+  sudo bash --login -eua $0 $1
```
pipefail here too

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+  exit 1
+fi
+
+# Register signal handling to clean the temporary directory in case of kill, kill -9, ctrl+c, error in the script
```
Good explanation, thanks.

Out of curiosity, why are you calling the function: `sighdl` is it a standard convention? what about a more self-explicative name? signal_handler?

francesco-giordano replied:
Not a standard convention. Changed

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+OS="$(grep "^ID=" /etc/os-release | cut -d"=" -f 2 | xargs)"
+
+case $OS in
+    ubuntu)
```
Let's use just 2 spaces for indentation.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+# Script to generate a system information archive.
+# This script needs to be run by a user in sudoers and it takes an existing directory as argument
+
+function body () {
```
I think we can improve the name of this function. What is it doing?

francesco-giordano replied:
Added comment on usage.

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,44 @@ def get_route_tables(subnet_id, region):
+
+
+def run_system_analyzer(scheduler_commands, remote_command_executor, request):
+    """Run script to collect system information on head and a compute node of a cluster"""
```
minor: add final dot.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,44 @@ def get_route_tables(subnet_id, region):
+    logging.debug(f"result.failed={result.failed}")
+    logging.debug(f"result.stdout={result.stdout}")
+    logging.info(
+        f"Copy results from remote cluster into: "
```
You don't need the `f` character in this line, because there are no variables to be replaced.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,44 @@ def get_route_tables(subnet_id, region):
+    scheduler_commands.wait_job_completed(job_id)
+    scheduler_commands.assert_job_succeeded(job_id)
+    logging.info(
+        f"Copy results from remote cluster into: "
```
same here

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,44 @@ def get_route_tables(subnet_id, region):
+    job_id = scheduler_commands.assert_job_submitted(result.stdout)
+    scheduler_commands.wait_job_completed(job_id)
+    scheduler_commands.assert_job_succeeded(job_id)
+    logging.info(
```
the lines from here to logging.info("Compute... can be moved into a `_retrieve_remote_report(node_type, report_dir)` common function.

Then you can call them with:
```python
_retrieve_remote_report("head_node", head_node_dir)
...
_retrieve_remote_report("compute_node", compute_node_shared_dir)
```

francesco-giordano replied:
Since head and compute use different method (`remote_command_executor.run_remote_script` and `scheduler_commands.submit_script`) to run the script, the function would be a duplicate of  `run_system_analyzer` or `remote_command_executor.get_remote_files`

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+function sighdl ()  {
+  echo "catching a signal!"
+    if [ "$1" != "0" ]; then
+      rm -fr "$BASE_TEMP_DIR"
```
here you're using an environment variable defined later, let's add a comment or let's pass it as parameter.

Then let's use `${VARIABLE_NAME}` syntax everywhere, for each bash variable.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+
+function _save_installed_services_timers() {
+  # Save installed services
+  systemctl --all --type=service --state running,active | head -n -7 | body sort -du > "${TEMP_DIR}"/services_active
```
same here we're using TEMP_DIR that is defined later, but what happens if it's not defined?

Let's pass it as parameter and then use it with `out_dir=$1` in this way the function always work as expected without relying on a variable that "should" be created later in the script or in the main function.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+
+function _save_scheduled_commands() {
+  # Save scheduled commands and scripts
+  mkdir "${TEMP_DIR}"/etc_cron
```
same comment here for TEMP_DIR and in the other functions.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+  [ "${EUID:-$(id -u)}" -eq 0 ]
+}
+
+# Check number of arguments
```
What about putting everything else into a main function?
```bash
main () {
...
}

main "$@"
```

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,169 @@
+
+_save_avail_mpi
+
+_save_network
```
network_info ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,222 @@
+}
+
+function body() {
+  # Function to print the heard passed in pipe and execute the argument
```
typo: header

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,222 @@
+
+function _save_imds_info() {
+  log "Save IMDSv2 information"
+  # Mics IMDS info
```
Misc

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/data/system-analyzer.sh:
```diff
@@ -0,0 +1,222 @@
+  log "Save mpi versions"
+  local OUT_DIR="${1}/mpi/"
+
+  # Save available mpi on server
```
"on server" can be removed

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/tests/common/utils.py:
```diff
@@ -231,3 +235,50 @@ def get_route_tables(subnet_id, region):
+    compute_node_shared_dir = "/opt/parallelcluster/shared"
+    head_node_dir = "/tmp"
+
+    logging.info(f"Creating remote_command_executor and scheduler_commands")
```
useless f-string.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,58 @@ Scheduling:
+
+In order to add the system analysis to a test do the following:
+1. Import the function from utils.py (e.g. `from tests.common.utils import run_system_analyzer`)
+2. Call the function after the cluster creation; can be useful to run it after all the other tests.
```
`after all the other tests.` --> as the last step of the test.

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,58 @@ Scheduling:
+In case of performance degradation detected in tests, to speed up the root cause analysis it is useful to have a
+report of all the services, cron settings and information about the tested cluster. The function `run_system_analyzer` creates a subfolder
+in `out_dir` called `system_analyzer`. Every run the function creates 2 files, one for the head node and one
+for a compute node of the fleet. The file generated is a `tar.gz` archive containing a directory structure
```
generated file

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,58 @@ Scheduling:
+In order to add the system analysis to a test do the following:
+1. Import the function from utils.py (e.g. `from tests.common.utils import run_system_analyzer`)
+2. Call the function after the cluster creation; can be useful to run it after all the other tests.
+If it is needed it is possible to specify the `partition` where the compute node have to be chosen
```
If needed, it is possible to specify the `partition` from which to collect compute node info. ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
````diff
@@ -976,3 +976,58 @@ Scheduling:
+```
+### How to compare system analysis
+
+The nodeJS `diff2html` generates a html file from a diff which helps in compare the differences.
````
helps to

francesco-giordano replied:
Done. Thanks

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,58 @@ Scheduling:
+
+The nodeJS `diff2html` generates a html file from a diff which helps in compare the differences.
+Compare result from different node roles (head, compute) can create misleading results: it is suggested to compare the
+same node role (e.g. head with head) 
```
node role --> node type

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
```diff
@@ -976,3 +976,58 @@ Scheduling:
+The nodeJS `diff2html` generates a html file from a diff which helps in compare the differences.
+Compare result from different node roles (head, compute) can create misleading results: it is suggested to compare the
+same node role (e.g. head with head) 
+Below an example on how compare system analysis:
```
system analysis results ?

francesco-giordano replied:
Done

enrico-usai commented on tests/integration-tests/README.md:
````diff
@@ -976,3 +976,58 @@ Scheduling:
+
+# Compare and generate the diff.html file
+diff --exclude=network/ -u *-run/system-information | diff2html -s side -i stdin -o stdout > diff.html ;
+```
+
+Once generate the file it can be inspected in the browser.
````
Once generated

francesco-giordano replied:
Done

francesco-giordano committed:
Add run_system_analyzer method to common utils and add system-analyzer.sh script

run_system_analyzer runs system-analyzer.sh script on head and a compute node to collect system information and copy them into output_dir
Add gathering_system_report in benchmark tests
Add Ubuntu support, kernel version and network statistics to system-information
Add support to partition in run_system_analyzer
Add run_system_analyzer to test_efa in case of OSU benchmark and p4d
